### ISSUE TITLE
Fix max col norm formula

### ISSUE THREAD

**Issue by goodfeli** [lisa-lab/pylearn2#1126]
Currently max col norm / kernel norm stuff uses (variations on) this formula:

updates[W] = updated_W * (desired_norms / (1e-7 + col_norms))

Adding 1e-7 causes the parameters to shrink gradually. This is usually not noticeable but it is very noticeable if the weights are small and you set the learning rate to 0 for debugging purposes. This could be fixed by using T.switch to set the scale factor to 1 for columns that are within the allowed norm range.

I think @lamblin is going to make a separate ticket to unify the max col norm implementations with a single ModelExtension plugin, so it could make sense to do that first so there is only one piece of code to fix.

**Comment by lamblin** [comment]
The refactoring of max col norm is not happening soon, so let's not wait to address this one.

**Comment by robintibor** [comment]
One question regarding this.
I saw in the pylearn terminal output during trianing that the actual max col norm for my softmax layer was exceeding the max col norm constraint that i specified (actual max col norm typically between 0.6 and 1.0, once even at 1.7, specified max_col_norm was 0.5) .
Is it intended that it's
```python 
col_norms = T.sqrt(T.square(W).sum(axis=0)) 
```
in https://github.com/lisa-lab/pylearn2/blob/master/pylearn2/model_extensions/norm_constraint.py#L47

and not 
```python 
col_norms = T.sqrt(T.square(updated_W).sum(axis=0)) 
```
as in https://github.com/lisa-lab/pylearn2/blob/master/pylearn2/models/mlp.py#L2045 ?

If it is correct this way, can I just ignore the output on the terminal saying the max col norm is too high? Hope I am not missing something here :)

**Comment by lamblin** [comment]
Using the norm of `updated_W` would make more sense to me.
@goodfeli can you confirm?

**Comment by goodfeli** [comment]
Yes, it should be updated_W, sorry.

2015-04-23 12:00 GMT-07:00 Pascal Lamblin <notifications@github.com>:

**Comment by lamblin** [comment]
Thanks.
@laurent-dinh can you take care of that as well?

**Comment by goodfeli** [comment]
I already sent a PR for the updated_W bug:
https://github.com/lisa-lab/pylearn2/pull/1498

2015-04-23 12:14 GMT-07:00 Pascal Lamblin <notifications@github.com>:

**Comment by lamblin** [comment]
Fixed in #1484.
